### PR TITLE
DPE-915 fix storage re-use bug

### PR DESCRIPTION
**Pull request by paulomach**
## Issue

Failing to reuse previously storage on the charm.

## Solution

If data directory is populated with mysql typical data, the charm tries to use start mysqld and rejoin the unit to the cluster.

_Note 1: The `unit-configured` unit peer key, that indicate if a unit was already configured, was replaced with a method to check the mysql data directory directly._ 

_Note 2: case there's a `clusteradmin` password rotation before the scale up, joining the unit will fail, since the passwords will be out of sync. There's a backlog task (dpe-2186) to track this scenario._

### CODE REVIEW

File: src/charm.py [hunk 1]
```diff
@@ -568,22 +574,24 @@ def _on_mysql_pebble_ready(self, event) -> None:
...
             self._reconcile_pebble_layer(container)
             return
 
+        self.unit.status = MaintenanceStatus("Initialising mysqld")
+
         # First run setup
         if not self._configure_instance(container):
-            self.unit.status = BlockedStatus("Unable to configure instance")
-            return
+            raise
```
[the remark this answers is not in the document]
Comment (paulomach) [reply src/charm.py:594]: The meaningful exception logs are done in the lib, here would just generate more noise.

File: src/mysql_k8s_helpers.py [hunk 2]
```diff
@@ -640,6 +640,19 @@ def start_mysqld(self) -> None:
...
+        stop_gr_command = (
+            f"shell.connect('{self.server_config_user}:{self.server_config_password}@{self.instance_address}')",
+            "data = session.run_sql('SELECT 1 FROM performance_schema.replication_group_members')",
+            "if len(data.fetch_all()) > 0:",
+            "    session.run_sql('STOP GROUP_REPLICATION')",
+        )
+        try:
+            self._run_mysqlsh_script("\n".join(stop_gr_command))
+        except ExecError:
+            logger.debug("Failed to stop Group Replication for unit")
```
[the remark this answers is not in the document]
Comment (paulomach) [reply src/mysql_k8s_helpers.py:654]: Not for the current use case, but I consider changing when there's a more sensitive use case.

File: src/mysql_k8s_helpers.py [hunk 3]
```diff
@@ -828,3 +841,39 @@ def _get_total_memory(self) -> int:
...
+                "private_key.pem",
+                "public_key.pem",
+                "server-cert.pem",
+                "server-key.pem",
+                "sys",
+                "undo_001",
+                "undo_002",
+            }
+
+            return expected_content <= content_set
```
[the remark this answers is not in the document]
Comment (paulomach) [reply src/mysql_k8s_helpers.py:877]: I :heart: `<=`

File: charmcraft.yaml [hunk 4]
```diff
@@ -11,9 +11,12 @@ bases:
...
-      - rustc
-      - cargo
+    charm-binary-python-packages:
+      - boto3==1.26.117
+      - cryptography==38.0.2
+      - jsonschema==4.16.0
+      - lightkube==0.12.0
+      - ops >= 2.0.0
+      - tenacity==8.0.1
+      - urllib3==1.26.15
```
[the remark this answers is not in the document]
Comment (paulomach) [reply charmcraft.yaml:21]: ops! yes - testing env.

File: src/charm.py [hunk 5]
```diff
@@ -465,6 +469,13 @@ def _rescan_cluster(self) -> None:
     # Charm event handlers
     # =========================================================================
 
+    def _on_peer_relation_joined(self, _) -> None:
+        """Handle the peer relation joined event."""
+        # set some initial unit data
+        self.unit_peer_data.setdefault("instance-hostname", self._get_unit_fqdn(self.unit.name))
```
[the remark this answers is not in the document]
Comment (paulomach) [reply src/charm.py]: 8926f83

File: charmcraft.yaml (hunk 4 above)
[the remark this answers is not in the document]
Comment (paulomach) [reply charmcraft.yaml]: 8926f83

**Review by taurus-forever — APPROVED** [review]
Nice progress! Let's the party begin!

**Pull request merged**